builtin/game/item_entity.lua

```diff
@@ -152,6 +177,104 @@ core.register_entity(":__builtin:item", {
 			return
 		end
 
+		local node_this = core.get_node_or_nil(pos)
+		local is_stuck = false
+		if node_this then
+			local sdef = minetest.registered_nodes[node_this.name]
+			is_stuck = (sdef.walkable == nil or sdef.walkable == true)
+				and (sdef.collision_box == nil or sdef.collision_box.type == "regular")
+				and (sdef.node_box == nil or sdef.node_box.type == "regular")
+		end
+
+		-- Push item out when stuck inside solid node
+		if is_stuck then
+			local shootdir
+			local cx = pos.x % 1
+			local cz = pos.z % 1
+			local order = {}
+
+			-- First prepare the order in which the 4 sides are to be checked.
+			-- 1st: closest
+			-- 2nd: other direction
+			-- 3rd and 4th: other axis
+			local cxcz = function(o, cw, one, zero)
+				if cw > 0 then
+					table.insert(o, { [one]=1, y=0, [zero]=0 })
+					table.insert(o, { [one]=-1, y=0, [zero]=0 })
+				else
+					table.insert(o, { [one]=-1, y=0, [zero]=0 })
+					table.insert(o, { [one]=1, y=0, [zero]=0 })
+				end
+				return o
+			end
+			if math.abs(cx) > math.abs(cz) then
+				order = cxcz(order, cx, "x", "z")
+				order = cxcz(order, cz, "z", "x")
+			else
+				order = cxcz(order, cz, "z", "x")
+				order = cxcz(order, cx, "x", "z")
+			end
+
+			-- Check which one of the 4 sides is free
+			for o=1, #order do
+				local nn = minetest.get_node(vector.add(pos, order[o])).name
+				local def_side = minetest.registered_nodes[nn]
+				if def_side and def_side.walkable == false and nn ~= "ignore" then
+					shootdir = order[o]
+					break
+				end
+			end
+			-- If none of the 4 sides is free, shoot upwards
+			if shootdir == nil then
+				shootdir = { x=0, y=1, z=0 }
+				local nn = minetest.get_node(vector.add(pos, shootdir)).name
+				if nn == "ignore" then
+					-- Do not push into ignore
+					return
+				end
+			end
+
+			-- Set new item moving speed accordingly
+			local newv = vector.multiply(shootdir, 3)
+			self.object:set_acceleration({x = 0, y = 0, z = 0})
+			self.object:set_velocity(newv)
+
+			disable_physics(self.object, self)
+
+			if shootdir.y == 0 then
+				self.force_out = newv
+				self.force_out_start = vector.floor(pos)
+				self.force_out_timer = 1
+			end
+			return
+		end
+
+		-- This code is run after the entity got a push from above “push away” code.
+		-- It is responsible for making sure the entity is entirely outside the solid node
+		-- (with its full collision box), not just its center.
+		if self.force_out_timer > 0 then
+			local cbox = self.object:get_properties().collisionbox
+			local ok = false
+			if self.force_out.x > 0 and (pos.x > (self.force_out_start.x + 0.5 + (cbox[4] - cbox[1])/2)) then ok = true
```

Comment:
From https://dev.minetest.net/Lua_code_style_guidelines#Misc:
>      You can put conditionals / loops with small conditions and bodies on one line. This is discouraged for all but the smallest ones though. 
>
> Good:
>
> ```lua
> local f, err = io.open(filename, "r")
> if not f then return err end
> 
> if     foo then return foo
> elseif bar then return bar
> elseif qux then return qux
> end
> ```
>
> Bad:
>
> ```lua
> if not f and use_error then error(err) elseif not f then return err end
> ```

Reply:
I'm surprised that is allowed, i'd prefer it wasn't done as it is much less readable.
I think the code style guidelines should be changed.

So, please don't do this here.

Reply:
I agree. The C++ code style guidelines is less lenient in this regard: https://dev.minetest.net/Code_style_guidelines#if_statements